EC J-PAKE: partial fix for role vs user+peer

mprse opened the pull request:
## Description

Apply a partial fix that allows applications following the spec to work in the special case where the user/peer IDs are "client"/"server" (in either order) instead setting role.

Resolves https://github.com/Mbed-TLS/mbedtls/issues/7149

This fix requires https://github.com/Mbed-TLS/mbedtls/pull/6567 and is created on top of this #6567 (all commits squashed into one commit). The changelog for #6567 will be included here.


## Gatekeeper checklist

- [x] **changelog** must also cover #6567 
- [x] **backport** not required
- [x] **tests** provided

mprse commented:
Only code style check fails on CI. Will be fixed after review/rebase.

mpg reviewed (changes requested):
Looks pretty good to me, only minor feedback mostly on tests.

One more point though: the description says this fixes #7149 but it's not addressing the last bullet point: make `psa_pake_set_role()` always return `PSA_ERROR_INVALID_ARGUMENT`, which it should, according to the spec. Is that on purpose? (For example, to facilitate the migration of existing users.)

mprse commented:
> One more point though: the description says this fixes https://github.com/Mbed-TLS/mbedtls/issues/7149 but it's not addressing the last bullet point: make psa_pake_set_role() always return PSA_ERROR_INVALID_ARGUMENT, which it should, according to the spec. Is that on purpose? (For example, to facilitate the migration of existing users.)

Yes, I wanted to ask about this. Why `PSA_ERROR_INVALID_ARGUMENT` should be returned. `PSA_ERROR_NOT_SUPPORTED` seems more suitable from my perspective. Argument can be valid, but operation is not supported by us.

mpg commented:
It's not that we don't support the operation, it's that the algorithm JPAKE doesn't support the operation. See https://arm-software.github.io/psa-api/crypto/1.1/ext-pake/api/pake.html#c.psa_pake_set_role and "J-PAKE does not assign roles to the participants, so it is not necessary to call `psa_pake_set_role()`." in the [the description or J-PAKE](https://arm-software.github.io/psa-api/crypto/1.1/ext-pake/api/pake.html#c.PSA_ALG_JPAKE).

OTOH, reading again the description of `psa_pake_set_role()`, it says "role is not a valid PAKE role, or is not supported for the operation’s algorithm." for `PSA_ERROR_NOT_SUPPORTED`, so I think it would be acceptable as well even though I think `INVALID_ARGUMENT` would be more precise in this case. (`NOT_SUPPORTED` suggests that another compliant implementation might accept the value, while in this case no compliant implementation should accept it.)

Note: we should keep accepting `ROLE_NONE`.

mprse commented:
On the second thought maybe we should check algorithm and for j-pake return invalid argument for roles different than `ROLE_NONE` (for `ROLE_NONE` just return success). For other algorithms return not supported?

mpg commented:
> On the second thought maybe we should check algorithm and for j-pake return invalid argument for roles different than `ROLE_NONE` (for `ROLE_NONE` just return success). For other algorithms return not supported?

Sounds good. Currently there are no other algorithms, but "not supported" still feels right - we are indeed not supporting algorithms before they are added to at least a draft of the spec.

mprse commented on include/psa/crypto_extra.h:
```diff
@@ -434,6 +434,9 @@ psa_status_t mbedtls_psa_inject_entropy(const uint8_t *seed,
 #define PSA_PAKE_OPERATION_STAGE_COLLECT_INPUTS 1
 #define PSA_PAKE_OPERATION_STAGE_COMPUTATION 2
 
+/** JPAKE operation stages. */
```
It is temporary, but I think it is useful to have it for now. Were do you suggest to move it?

mpg replied:
I'd say being temporary is an additional reason not to be in a public header :) I'm not sure where to move it, though.

I think `ssl_tls.c` should have its own copy: it's using those strings because there's a TLS RFC (well, draft) that says so, regardless of what the PSA implementation supports or not.

For the other users (`psa_crypto.c`, `test_suite_psa_crypto_driver_wrappers.function` and `test_suite_psa_crypto_pake.function`) I'm not sure. I'm tempted to say each of them could have its own copy: it's not like we're going to change those values, and we already have copies of those strings in `test_suite_psa_crypo_pake.data` anyway.

Wdyt?

mprse replied:
Ok I'll move this to source files.

mprse commented on tests/suites/test_suite_psa_crypto_pake.data:
```diff
@@ -1,90 +1,106 @@
 PSA PAKE: set invalid user
 depends_on:PSA_WANT_KEY_TYPE_ECC_KEY_PAIR:PSA_WANT_ECC_SECP_R1_256:PSA_WANT_ALG_SHA_256
-ecjpake_setup:PSA_ALG_JPAKE:PSA_KEY_TYPE_PASSWORD:PSA_KEY_USAGE_DERIVE:PSA_PAKE_PRIMITIVE(PSA_PAKE_PRIMITIVE_TYPE_ECC, PSA_ECC_FAMILY_SECP_R1, 256):PSA_ALG_SHA_256:PSA_PAKE_ROLE_SERVER:0:ERR_INJECT_INVALID_USER:PSA_ERROR_INVALID_ARGUMENT
+ecjpake_setup:PSA_ALG_JPAKE:PSA_KEY_TYPE_PASSWORD:PSA_KEY_USAGE_DERIVE:PSA_PAKE_PRIMITIVE(PSA_PAKE_PRIMITIVE_TYPE_ECC, PSA_ECC_FAMILY_SECP_R1, 256):PSA_ALG_SHA_256:"aaaa":"server":0:ERR_INJECT_SET_USER:PSA_ERROR_NOT_SUPPORTED
```
Good point. I changed the idea and provided peer/user values via arguments and forgot to change errors.

mpg reviewed (dismissed):
LGTM, thanks for addressing my comments!

mprse commented:
Rebased, fixed `unused variable CI failure`, added change log for driver dispatch for EC J-PAKE (was missing in https://github.com/Mbed-TLS/mbedtls/pull/6567) and fixed some strange mismatch in `set_peer` and `set_user` functions that appeared after rebase.

mpg commented on ChangeLog.d/fix-jpake-user-peer.txt:
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Allow setting user and peer identity for EC J-PAKE operation
+     instead role in PAKE PSA Crypto API.
```
Perhaps add: "as described in the specification" to make it clear that it was a compliance bug.

mprse replied:
I will wait for the CI result and then fix it with other potential problems.

mprse replied:
CI was green. Force pushed fixed change log.

mpg reviewed (dismissed):
Looks good to me, just suggested a minor improvement to the ChangeLog entry.

mprse commented:
Previous CI run (before updating the change log) was fine. Now I see that few checks are marked as aborted. Probably re-run is needed.

It would be good to merge this one till Wednesday, because at the end of the week I will be off and will not have access to laptop.

mpg commented:
Internal CI's pr-head is fully green, so I don't think we need to restart OpenCI for the parts that were aborted. The results of ABI-API-checking are as expected: the ABI is modified as some structures are changed, and that's OK.

mpg reviewed (dismissed):
LGTM. Thanks!

ronald-cron-arm commented on include/psa/crypto_extra.h:
```diff
@@ -1344,6 +1344,70 @@ psa_status_t psa_crypto_driver_pake_get_role(
     const psa_crypto_driver_pake_inputs_t *inputs,
     psa_pake_role_t *role);
 
+/** Get the lengths of the user id in bytes from given inputs.
```
I may not remember all the history of discussions here but in #5451 which has been a reference to define the PAKE driver interface, there are two ways for the driver to get a buffer of early data.
Through two APIs:
```C
psa_status_t psa_crypto_driver_key_derivation_get_input_size(
    const psa_crypto_driver_key_derivation_inputs_t *inputs,
    psa_key_derivation_step_t step,
    size_t *size);
psa_status_t psa_crypto_driver_key_derivation_get_input_bytes(
    const psa_crypto_driver_key_derivation_inputs_t *inputs,
    psa_key_derivation_step_t step,
    uint8_t *buffer, size_t buffer_size, size_t *buffer_length);
```
Through only one:
```C
psa_status_t psa_crypto_driver_key_derivation_get_input_key(
    const psa_crypto_driver_key_derivation_inputs_t *inputs,
    psa_key_derivation_step_t step,
    uint8_t** p_key_buffer, size_t *key_buffer_size);
```
With the second one you give a chance to the driver to avoid a memory allocation. And as far as I remember the first way was introduced in the case we need to generate on-the-fly data in the core (but again I may not remember well or all the reasons).
Here I would say that the second way (only one function) is better suited. I know that for the password we've gone the first way but I am not sure now it was the right call (no need to change it in this PR though).

ronald-cron-arm replied:
> We are using the first way as it is designed for for transparent drivers and the second form is for opaque drivers

Why do you say so?

mprse replied:
Please see https://github.com/Mbed-TLS/mbedtls/pull/6567#discussion_r1082953367.

ronald-cron-arm replied:
I am not convinced yet that we can infer that the first way is designed for transparent drivers and the second for opaque drivers. Anyway, better for the get password/peer/user to be aligned thus let's keep it that and I am creating an issue to decide if we keep things as they are or if we move to the second way.

ronald-cron-arm commented on ChangeLog.d/fix-jpake-user-peer.txt:
```diff
@@ -0,0 +1,4 @@
+Bugfix
+   * Allow setting user and peer identity for EC J-PAKE operation
+     instead role in PAKE PSA Crypto API as described in the specification.
+     This is partial fix that allows only "client"/"server" ids.
```
```suggestion
   * Allow setting user and peer identifiers for EC J-PAKE operation
     instead of role in PAKE PSA Crypto API as described in the specification.
     This is a partial fix that allows only "client" and "server" identifiers.
```

ronald-cron-arm commented on library/psa_crypto.c:
```diff
@@ -7458,14 +7566,25 @@ static psa_status_t psa_pake_complete_inputs(
-        inputs.role != PSA_PAKE_ROLE_CLIENT &&
-        inputs.role != PSA_PAKE_ROLE_SERVER) {
-        return PSA_ERROR_NOT_SUPPORTED;
+    if (operation->alg == PSA_ALG_JPAKE) {
```
This logic rather belongs to `psa_crypto_pake.c` to me as if I understand correctly this is to adapt to the interface of ecjpake.c.  That way we would also already add at the right place the calls to `get_user/peer_id` in `mbedtls_psa_pake_setup()`.

mprse replied:
Yes, this could be handled in `psa_crypto_pake.c`, but as this is only partial fix I wanted to handle everything at core level. This part will be removed in follow-up PR when driver part will be adapted to use `peer/user`.

mpg replied:
I agree with @mprse that the goal of this PR is only to fix the user-facing API; we'll need a follow-up for drivers where we can fix this. We've got only two days to get this into the release, and the slow CI doesn't exactly help iterating quickly.

ronald-cron-arm replied:
okay let's keep as it is.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -5,6 +5,11 @@
 int pake_in_driver = 0;
+
+/* JPAKE user/peer ids. */
+#define JPAKE_SERVER_ID "server"
+#define JPAKE_CLIENT_ID "client"
```
ditto, const uint8_t ...

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -50,6 +53,10 @@ typedef enum {
 } pake_round_t;
 
+/* JPAKE user/peer ids. */
+#define JPAKE_SERVER_ID "server"
+#define JPAKE_CLIENT_ID "client"
```
ditto, const uint8_t ...

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -537,7 +544,7 @@ exit:
 /* BEGIN_CASE depends_on:PSA_WANT_ALG_JPAKE */
 void ecjpake_setup(int alg_arg, int key_type_pw_arg, int key_usage_pw_arg,
-                   int primitive_arg, int hash_arg, int role_arg,
+                   int primitive_arg, int hash_arg, char *user_arg, char *peer_arg,
```
```suggestion
                   int primitive_arg, int hash_arg, char *user_arg, char *peer_arg,
```
use rather data_t.

mprse replied:
With `data_t` I would have to provide bytes in hex as input. Instead `"client"` it would be `"636C69656E74"`. I think null-terminated string is better for this case.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -724,6 +737,10 @@ void ecjpake_rounds_inject(int alg_arg, int primitive_arg, int hash_arg,
     ecjpake_error_stage_t err_stage = err_stage_arg;
+    const uint8_t server_id[] = JPAKE_SERVER_ID;
+    const uint8_t client_id[] = JPAKE_CLIENT_ID;
+    const size_t server_id_len = strlen(JPAKE_SERVER_ID);
+    const size_t client_id_len = strlen(JPAKE_CLIENT_ID);
```
```suggestion
    const size_t server_id_len = ARRAY_LENGTH(server_id);
    const size_t client_id_len = ARRAY_LENGTH(client_id);
```

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -1030,8 +1055,9 @@ void pake_input_getters_role()
-    PSA_ASSERT(psa_pake_set_role(&operation, PSA_PAKE_ROLE_SERVER));
-
+    /* Role can not be set directly using psa_pake_set_role(). It is set by the core
+       based on given user/peer. Simulate that Role is already set. */
```
```suggestion
       based on given user/peer identifiers. Simulate that Role is already set. */
```

mprse commented on library/psa_crypto.c:
```diff
@@ -91,6 +91,10 @@
 #endif
 
+/* JPAKE user/peer ids. */
+#define JPAKE_SERVER_ID "server"
+#define JPAKE_CLIENT_ID "client"
```
I was considering this, but I couldn't find such array of bytes in mbedtls code, so I have chosen to use simple macro as this is only temporary solution.

mprse replied:
While building with drivers I got the following linking error:
```
/usr/bin/ld: ../library/libtestdriver1.a(psa_crypto.o):(.rodata+0x0): multiple definition of `jpake_server_id'; ../library/libmbedcrypto.a(psa_crypto.o):(.rodata+0x0): first defined here
/usr/bin/ld: ../library/libtestdriver1.a(psa_crypto.o):(.rodata+0x6): multiple definition of `jpake_client_id'; ../library/libmbedcrypto.a(psa_crypto.o):(.rodata+0x6): first defined here
```

Seems like a problem with copy for the driver.

ronald-cron-arm replied:
With `static const uint8_t ...` it should be ok.

mprse replied:
Yes `static` solves this problem and its a valid solution. Also checked that changing name to `psa_xxxx` also helps, but it is not right.

mprse commented on library/psa_crypto.c:
```diff
@@ -7477,6 +7596,12 @@ static psa_status_t psa_pake_complete_inputs(
+    /* User and peer are translated to role. */
+    mbedtls_free(inputs.user);
+    mbedtls_free(inputs.peer);
+    inputs.user = NULL; inputs.user_len = 0;
+    inputs.peer = NULL; inputs.peer_len = 0;
```
Good point. This is just a copy on stack and real `inputs` are already zeroed.

mprse commented:
CI is green. Ready for next review round.

mpg commented on library/psa_crypto.c:
```diff
@@ -91,9 +91,9 @@
-#define JPAKE_CLIENT_ID "client"
+/* The only two JPAKE user/peer identifiers supported for the time being. */
+static const uint8_t jpake_server_id[] = { 's', 'e', 'r', 'v', 'e', 'r' };
+static const uint8_t jpake_client_id[] = { 'c', 'l', 'i', 'e', 'n', 't' };
```
Off-topic, but spelling it out like this reminds me of adults talking about c, h, o, c, o, l, a, t, e in front of toddlers so that they won't hear the word, which made me smile :)

mpg commented on library/ssl_tls.c:
```diff
@@ -1953,15 +1949,19 @@ void mbedtls_ssl_set_verify(mbedtls_ssl_context *ssl,
 #if defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED)
 
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
+/* The only two JPAKE user/peer identifiers supported for the time being. */
```
When it comes to TLS, that comment is not accurate. That's the two strings specified by https://datatracker.ietf.org/doc/html/draft-cragie-tls-ecjpake-01 and as such the only we'll ever use here.

mpg reviewed (dismissed):
Looks good to me. Just one comment to fix, but if it's the only thing, it can be done in a follow-up to avoid an extra round of CI+review.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -1041,3 +1059,119 @@ exit:
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:PSA_WANT_ALG_JPAKE:PSA_ALG_SHA_256 */
```
```suggestion
/* BEGIN_CASE depends_on:PSA_WANT_ALG_JPAKE */
```
does not seem to depend on SHA-256 as we do not do any crypto calculation here.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -1041,3 +1059,119 @@ exit:
+    psa_pake_cipher_suite_t cipher_suite = psa_pake_cipher_suite_init();
+    psa_pake_operation_t operation = psa_pake_operation_init();
+    const uint8_t peer[] = "server";
+    const size_t peer_len = strlen("server");
```
```suggestion
    const size_t peer_len = ARRAY_LENGTH(peer);
```

ronald-cron-arm reviewed (changes requested):
A few final things on my side, mostly on documentation, some improvement suggestions but also some copy/paste issues. Otherwise this looks good to me.

mprse commented on tests/suites/test_suite_psa_crypto_pake.function:
```diff
@@ -1041,3 +1059,119 @@ exit:
+    psa_pake_cipher_suite_t cipher_suite = psa_pake_cipher_suite_init();
+    psa_pake_operation_t operation = psa_pake_operation_init();
+    const uint8_t user[] = "server";
+    const size_t user_len = strlen("server");
```
`ARRAY_LENGTH` won't work in this case as it will include `null` char for string termination. This is why `strlen` is used.
I will use the following version if `strlen` is a problem:
```
    const uint8_t peer[] = { 's', 'e', 'r', 'v', 'e', 'r' };
    const size_t peer_len = sizeof(peer);
```

mpg reviewed (approved):
LGTM

ronald-cron-arm reviewed (approved):
LGTM, thanks.

mprse commented:
Ready for merge.
